### PR TITLE
Support for panel customisation in activity files

### PR DESCRIPTION
**Pull request by barnettwilliam**
Added support for customising panel icon and buttons from the activity configuration for #102 and added Button class to improve cohesion.

Updated documentation located mdenet/educationplatform.wiki@e788820f37e2715d2c703504779cf3560983c56b
on branch `102-add-support-for-panel-customisation-in-activity-files`

### CODE REVIEW

File: platform/src/OutputPanel.js [hunk 1]
```diff
@@ -1,5 +1,7 @@
+
 import { ModelPanel } from "./ModelPanel.js";
-import { language } from "./Playground.js";
+import { language } from "./Playground.js"
```
Comment (szschaler) [inline platform/src/OutputPanel.js:3]: Why did you remove the semicolon here?

File: platform/src/Panel.js [hunk 2]
```diff
@@ -97,27 +97,14 @@ class Panel {
...
 
         var buttonViewData= buttons.map( (btn) => {
-            var buttonData={};
-            
-            buttonData.html= this.buttonHtml(btn.icon, btn.hint);
-            buttonData.cls= "sys-button";
-            buttonData.onclick= btn.action;
-
-            return buttonData;
+            return btn.getView();
```
Comment (szschaler) [inline platform/src/Panel.js:107]: But elsewhere you're already calling `.getView()` when defining the buttons. Are these consistent?

File: platform/src/Playground.js [hunk 3]
```diff
@@ -262,8 +263,8 @@ function initialisePanels() {
    /**
      * Create a panel for a given panel config entry
      * 
-     * @param {string} panel 
-     * @return {Panel}
+     * @param {Object} activity config panel definition 
+     * @return {Panel} platform Panel
      */
     function createPanelForDefinitionId(panel){
```
Comment (szschaler) [inline platform/src/Playground.js:269]: Is the parameter name `panel` still correct?

File: platform/src/ConsolePanel.js [hunk 4]
```diff
@@ -13,11 +14,15 @@ class ConsolePanel extends Panel {
     }
 
     getButtons() {
```
Comment (szschaler) [inline platform/src/ConsolePanel.js:16]: Where is this method used?

**Review by szschaler — COMMENTED** [review]
Generally looks good except for some bits that I didn't understand...

File: platform/src/ConsolePanel.js (hunk 4 above)
Comment (barnettwilliam) [reply platform/src/ConsolePanel.js:16]: It's called in the ConsolePanel constructor on line 11 which sets the member variable that gets used to render the buttons.

Comment (szschaler) [reply platform/src/ConsolePanel.js:16]: Panel defines an `addButtons` method. Why is that not used in the `ConsolePanel` constructor?

**Comment by szschaler** [comment]
I guess what confuses me is that we seem to have three different mechanisms for defining buttons with this PR. They appear to all be subtly different, but I may also just be misinterpreting...

File: platform/src/Playground.js (hunk 3 above)
Comment (barnettwilliam) [reply platform/src/Playground.js:269]: The parameter name is as intended the updated description removed the parameter name.
I'll update it to `@param {Object} panel -  The activity config panel definition.`

The jsdoc comment style supports dashes to separate paramter name from description which make this clearer raised #107 to apply consistently.

File: platform/src/Panel.js (hunk 2 above)
Comment (barnettwilliam) [reply platform/src/Panel.js:107]: There are two ways Buttons are intialised
- for built-in panels that aren't customisable they are intialised in the panel constructor with a call to the panel's getButtons() method e.g. the ConsolePanel line 11.

- for panels that are customisable like those defined by tools in the configuration they use the Panel super class'  addButtons() method called during platform intialisation in Playground.js line 352.

With the added customisation of panels that override those defined in tool panel def there are likely to be odd behaviours introduced e.g. overriding a non-customisable panel's button.

File: platform/src/OutputPanel.js (hunk 1 above)
Comment (barnettwilliam) [reply platform/src/OutputPanel.js:3]: Agree this was unintentional so will revert.

File: platform/src/Panel.js (hunk 2 above)
Comment (szschaler) [reply platform/src/Panel.js:107]: But why are those uses not consistent? One uses `getView()` in the definition of the button, the other uses it when adding the button via `addButtons()`.

How would buttons be overwritten and what would happen in that case?

Comment (barnettwilliam) [reply platform/src/Panel.js:107]: I agree that the addButton() can be used in same way for the built-in panels so the getButtons() can be easily removed.

For overwriting the built-in panels buttons its likely you would be able to define buttons new buttons but not reference the panel defined ones by id. I will test this case though to see.

Comment (barnettwilliam) [reply platform/src/Panel.js:107]: For the case of trying to overwrite a built-in panel's buttons the the panels config would have no buttons declared since they are instantiated by the panel's js. Therefore panelDefinition.buttons is undefined so the buttons are not modified by lines 335 or 352 - this is the intended behaviour confirmed by test.

**Comment by barnettwilliam** [comment]
You were right to be confused, I have added the mentioned fixes in 1032f28d23a6bfd62a52818ba38a225ddbc0cf43
